scaleway/resource_iam_group.go

```diff
@@ -0,0 +1,221 @@
+			},
+			"user_ids": {
+				Type:        schema.TypeList,
+				Description: "List of IDs of the users attached to the group",
+				Optional:    true,
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					ValidateFunc: validationUUID(),
+				},
+			},
+			"application_ids": {
+				Type:        schema.TypeList,
+				Description: "List of IDs of the applications attached to the group",
+				Optional:    true,
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					ValidateFunc: validationUUID(),
+				},
+			},
+			"organization_id": organizationIDSchema(),
+		},
+	}
+}
+
+func resourceScalewayIamGroupCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	api := iamAPI(meta)
+	req := &iam.CreateGroupRequest{
+		OrganizationID: d.Get("organization_id").(string),
+		Name:           expandOrGenerateString(d.Get("name"), "group-"),
+		Description:    d.Get("description").(string),
+	}
+	group, err := api.CreateGroup(req, scw.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	appIds := d.Get("application_ids").([]interface{})
+	userIds := d.Get("user_ids").([]interface{})
+	if len(appIds) > 0 || len(userIds) > 0 {
+		appIdsStr := []string(nil)
+		for _, id := range appIds {
+			appIdsStr = append(appIdsStr, id.(string))
+		}
+		userIdsStr := []string(nil)
+		for _, id := range userIds {
+			userIdsStr = append(userIdsStr, id.(string))
+		}
+		_, err := api.SetGroupMembers(&iam.SetGroupMembersRequest{
+			GroupID:        group.ID,
+			ApplicationIDs: appIdsStr,
+			UserIDs:        userIdsStr,
+		}, scw.WithContext(ctx))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	d.SetId(group.ID)
+
+	return resourceScalewayIamGroupRead(ctx, d, meta)
+}
+
+func resourceScalewayIamGroupRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	api := iamAPI(meta)
+	group, err := api.GetGroup(&iam.GetGroupRequest{
+		GroupID: d.Id(),
+	}, scw.WithContext(ctx))
+	if err != nil {
+		if is404Error(err) {
+			d.SetId("")
+			return nil
+		}
+		return diag.FromErr(err)
+	}
+
+	_ = d.Set("name", group.Name)
+	_ = d.Set("description", group.Description)
+	_ = d.Set("created_at", flattenTime(group.CreatedAt))
+	_ = d.Set("updated_at", flattenTime(group.UpdatedAt))
+	_ = d.Set("organization_id", group.OrganizationID)
+	_ = d.Set("user_ids", group.UserIDs)
+	_ = d.Set("application_ids", group.ApplicationIDs)
+
+	return nil
+}
+
+func resourceScalewayIamGroupUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	api := iamAPI(meta)
+
+	group, err := api.GetGroup(&iam.GetGroupRequest{
+		GroupID: d.Id(),
+	}, scw.WithContext(ctx))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	req := &iam.UpdateGroupRequest{
+		GroupID: group.ID,
+	}
+
+	if d.HasChange("name") {
+		req.Name = expandStringPtr(d.Get("name").(string))
+	} else {
+		req.Name = &group.Name
```

Comment:
Why should we have this else branch? Shouldn't we send only the thing we wish to modify?

Reply:
The Update request uses the PATCH method, not PUT so we have to send all information, even if we don't want to modify all of them. But indeed the way I wrote it could be improved so I've changed it a bit.